### PR TITLE
add pagination for keys in each lock

### LINKED ISSUE #465: Add pagination to keys for each lock on Dashboard
https://github.com/unlock-protocol/unlock/blob/5decabbfd2098ec96deb41717b27f77963fb2eca/unlock-app/src/components/creator/lock/KeyList.js#L7-L12

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in 5decabbfd2098ec96deb41717b27f77963fb2eca when #462 was merged. cc @unlock-protocol.

### PR DESCRIPTION
**Pull request by wawhal**
Fix #465 

gif

## Proposed Changes

Adding pagination for the list of keys for each lock
- 10 items per page
- Can go to previous and next page
- Can jump to first page or last page
- Can jump to a page directly
- This pagination component can be reused for pagination purposes

Things that I had proposed but I could not see them fitting in the design provided in the issue:
- Giving ability to set the number of items per page

I am new to this crypto world and don't understand much of these contracts and stuff. I am a React developer and I worked on the pagination bit by using dummy data.

Let me know if there is anything you would like to change :-)

### CODE REVIEW

**Comment by julien51** [comment]
Thanks a lot @wawhal ! I'll review all that ASAP. Do you mind adding a screenshot (or even the screen recording you added on the issue, this will allow for a better idea of what is going on here). Thanks!

File: unlock-app/src/__tests__/middlewares/currencyConversionMiddleware.test.js [hunk 1]
```diff
@@ -4,8 +4,12 @@ import { setConversionRate } from '../../actions/currencyconvert'
 afterEach(() => mockAxios.reset())
 
 describe('Currency conversion service retrieval middleware', () => {
```
Comment (julien51) [inline unlock-app/src/__tests__/middlewares/currencyConversionMiddleware.test.js]: Could you remove this file from your commit since these changes are not related to the purpose of your PR? Thanks!

File: unlock-app/src/components/creator/lock/KeyList.js [hunk 2]
```diff
@@ -3,26 +3,37 @@ import { connect } from 'react-redux'
...
 
 // TODO add pagination
 export function KeyList({ keys }) {
+  const renderItems = items => {
+    return items.map(item => {
+      return (
+        <Row key={item.id}>
+          <Data>{item.transaction}</Data>
+          <Cell>{expirationAsDate(item.expiration)}</Cell>
+          <Data>{item.key}</Data>
```
Comment (julien51) [inline unlock-app/src/components/creator/lock/KeyList.js]: that would be `item.data`. Also do you mind actually renaming the `item` variable into `key` as it is more explicit (or `lockKey` since `key` is also used by jsx)

File: unlock-app/src/components/interface/pagination/Next.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  }
+  const goToLastPage = () => {
+    if (currentPage + 1 > numberOfPages) {
+      return
+    }
+    goToPage(numberOfPages)
+  }
+  return (
+    <ArrowGroup>
+      <RightArrow onClick={goToNextPage}>&gt;</RightArrow>
```
Comment (julien51) [inline unlock-app/src/components/interface/pagination/Next.js]: I think it would be good to maybe have a `disabled` state when there are no next page?

File: unlock-app/src/components/interface/pagination/PageNumbers.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      <PageNumberActive onClick={() => goToPage(number)}>
+        {number}
+      </PageNumberActive>
+    ) : (
+      <PageNumber onClick={() => goToPage(number)}>{number}</PageNumber>
+    )
+  }
+
+  // If number of pages less than 10, show all 10 page numbers
+  if (numberOfPages <= 10) {
```
Comment (julien51) [inline unlock-app/src/components/interface/pagination/PageNumbers.js]: Maybe best to define `MAX_NUMBER_OF_PAGES_TO_SHOW_ALL`  constant in `constants.js`?

File: unlock-app/src/components/interface/pagination/PageNumbers.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+          })}
+      </PageGroup>
+    )
+  }
+
+  // If number of pages greater than 10,  show first five and last five (1 2 3 4 5 .... 26 27 28 29 30)
+  return (
+    <PageGroup>
+      {//show first five page numbers
+        Array(5)
```
Comment (julien51) [inline unlock-app/src/components/interface/pagination/PageNumbers.js]: Same here, please use a constant so it is really easy to change to another number should we need to!

File: unlock-app/src/components/interface/pagination/Pagination.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import styled from 'styled-components'
+import PreviousButtons from './Previous'
+import NextButtons from './Next'
+import PageNumbers from './PageNumbers'
+
+class Paginate extends React.Component {
+  constructor(props) {
+    super(props)
+    const itemCount = props.items.length,
+      itemsPerPage = 10
```
Comment (julien51) [inline unlock-app/src/components/interface/pagination/Pagination.js]: Another constant ;)

File: unlock-app/src/components/interface/pagination/Next.js (hunk 3 above)
Comment (wawhal) [reply unlock-app/src/components/interface/pagination/Next.js]: Not sure if that is possible. Because the button is a `div` and not a `button`. `div` doesn't take a `disabled` attribute.

File: unlock-app/src/__tests__/services/web3Service.test.js [hunk 7]
```diff
@@ -850,7 +850,10 @@ describe('Web3Service', () => {
       it('should emit a new transaction once it has been submitted', done => {
```
[the remark this answers is not in the document]
Comment (wawhal) [reply unlock-app/src/__tests__/services/web3Service.test.js]: These files are automatically being changed as I commit my changes. Some reformatter is doing that. As you see, I have not made any change to them. Also, the package-lock is exactly the same as that in master, still it shows diff, I have no clue why :(

**Comment by wawhal** [comment]
@julien51 I have worked on the changes that you proposed and pushed the updates. Updated the PR with the GIF.
I have also added stories for the PageNumber component as it has quite a lot of edge cases.
From the UI testing point of view, I have added a test case for the PageNumber component.  
Rest of the components are fairly declarative and I haven't added tests for them. Let me know if you need tests for them anyway. :-)

File: unlock-app/src/__tests__/services/web3Service.test.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (julien51) [reply unlock-app/src/__tests__/services/web3Service.test.js]: huh, this is strange. Ok thanks for trying anyway :(

File: unlock-app/src/components/interface/pagination/Next.js (hunk 3 above)
Comment (julien51) [reply unlock-app/src/components/interface/pagination/Next.js]: I understand but you could make its "style" show that it is disabled (lighter?) and with no click possible on it. I tend to think that if it looks the same but behave differently, this is not a great experience... 
I actually think it would be better to use `<a>` rather than div in order to remain semantic here?

File: unlock-app/src/components/creator/lock/KeyList.js [hunk 8]
```diff
@@ -5,15 +5,14 @@ import UnlockPropTypes from '../../../propTypes'
 import { expirationAsDate } from '../../../utils/durations'
 import Pagination from '../../interface/pagination/Pagination'
 
-// TODO add pagination
 export function KeyList({ keys }) {
-  const renderItems = items => {
-    return items.map(item => {
+  const renderItems = lockKeys => {
+    return lockKeys.map(lockKey => {
```
Comment (julien51) [inline unlock-app/src/components/creator/lock/KeyList.js]: nit: destructure? `return lockKeys.map(({id, transaction, expiration, data})=> {` and then use them directly below?

File: unlock-app/src/components/interface/pagination/Next.js (hunk 3 above)
Comment (wawhal) [reply unlock-app/src/components/interface/pagination/Next.js]: Makes sense. I'll make the changes.

**Comment by wawhal** [comment]
@julien51 I resolved all the reviews that you requested. Except the package-lock :-(
I don't know what to do with it. I reset it to master and pushed it, it still shows conflict.
Please give me specific instructions on how to resolve it, or please resolve it in the UI before merging.
Sorry :-(

**Comment by cellog** [comment]
@wawhal you can probably fix the package-lock.json problem by upgrading npm with

`npm i -G npm`

and then run npm i in the folder, it should revert package-lock.json back to original

**Comment by wawhal** [comment]
The tests seem to be working locally for me. Not sure why the build is failing.

**Comment by julien51** [comment]
Thanks a lot for your work on this @wawhal . Please give a little more time to review the PR.

File: unlock-app/src/__tests__/middlewares/currencyConversionMiddleware.test.js (hunk 1 above)
Comment (julien51) [reply unlock-app/src/__tests__/middlewares/currencyConversionMiddleware.test.js]: This should be fixed once https://github.com/unlock-protocol/unlock/pull/732 has been merged. Could you please rebase after that? Thanks!

File: unlock-app/src/__tests__/services/web3Service.test.js [hunk 9]
```diff
@@ -823,7 +823,10 @@ describe('Web3Service', () => {
...
         expect.assertions(2)
         web3Service.sendTransaction = jest.fn((transaction, args, cb) => {
-          return cb(null, { event: 'transactionHash', args: { hash: '0x123' } })
```
Comment (julien51) [inline unlock-app/src/__tests__/services/web3Service.test.js]: Same here. This should be fixed once https://github.com/unlock-protocol/unlock/pull/732 has been merged. Could you please rebase after that? Thanks!

File: unlock-app/src/__tests__/components/interface/PaginationPageNumbers.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import React from 'react'
+import * as rtl from 'react-testing-library'
+import 'jest-dom/extend-expect'
```
Comment (cellog) [inline unlock-app/src/__tests__/components/interface/PaginationPageNumbers.js]:
```suggested
```

This is loaded in the jest.setup.js

File: unlock-app/src/components/creator/CreatorLock.js [hunk 11]
```diff
@@ -78,7 +83,7 @@ export class CreatorLock extends React.Component {
         )}
         {!showEmbedCode &&
           showKeys && (
-          <LockPanel>
+          <LockPanel onClick={this.preventPropagation}>
```
Comment (cellog) [inline unlock-app/src/components/creator/CreatorLock.js]: Instead of using preventPropagation and binding it in the constructor, I recommend:

```js
<LockPanel onClick={e => e.preventPropagation}>
```

This is much simpler and achieves the same result

File: unlock-app/src/components/creator/lock/KeyList.js [hunk 12]
```diff
@@ -48,6 +58,10 @@ const mapStateToProps = (state, { lock }) => {
 
 export default connect(mapStateToProps)(KeyList)
 
+const KeyListWrapper = styled.div`
```
Comment (cellog) [inline unlock-app/src/components/creator/lock/KeyList.js]: instead of creating a new styled.div, perhaps you can put this inside the Table definition? I think this is OK because we don't export it.

File: unlock-app/src/components/interface/pagination/PageNumbers.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+  if (numberOfPages <= PGN_MAX_NUMBER_OF_PAGES_TO_SHOW_ALL) {
+    return (
+      <PageGroup>
+        {Array(numberOfPages)
+          .fill()
+          .map((_, pn) => {
+            return pageNumber(pn + 1)
+          })}
+      </PageGroup>
+    )
```
Comment (cellog) [inline unlock-app/src/components/interface/pagination/PageNumbers.js]: The code:

```js
Array(numberOfPages)
          .fill()
          .map((_, pn) => {
            return pageNumber(pn + 1)
          })
```

is used quite a bit in this component. Perhaps we can do this once, and use slices, so for example on line 41:

```js
{Allpages.slice(0, 5)}
```

and a similar usage on line 81 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice

File: unlock-app/src/components/interface/pagination/PageNumbers.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+  return (
+    <PageGroup>
+      {//show first five page numbers
+        Array(HALF_MAX_PAGES_TO_SHOW_ALL)
+          .fill()
+          .map((_, pn) => {
+            return pageNumber(pn + 1)
+          })}
+      {//show first few dots
+        currentPage - 1 === HALF_MAX_PAGES_TO_SHOW_ALL ? (
```
Comment (cellog) [inline unlock-app/src/components/interface/pagination/PageNumbers.js]: Could you extract this complex logic into some constants and then render them so it is easier to follow the render logic?

File: unlock-app/src/components/interface/pagination/Pagination.js [hunk 15]
```diff
@@ -0,0 +1,119 @@
+import React from 'react'
+import styled from 'styled-components'
+import PreviousButtons from './Previous'
+import NextButtons from './Next'
+import PageNumbers from './PageNumbers'
+import { PGN_ITEMS_PER_PAGE } from '../../../constants'
+
+class Paginate extends React.Component {
+  constructor(props) {
```
Comment (cellog) [inline unlock-app/src/components/interface/pagination/Pagination.js]: Looks great! can you add propTypes?

```js
     static propTypes  = {
      ...
     }
```

You should `import UnlockPropTypes from 'src/propTypes'` and then use them. Add any you don't find.

Don't forget to run `npm run lint`! It will catch these kinds of omissions

File: unlock-app/src/stories/interface/Pagination.stories.js [hunk 16]
```diff
@@ -0,0 +1,73 @@
+import React from 'react'
```
Comment (cellog) [inline unlock-app/src/stories/interface/Pagination.stories.js]: fabulous!

**Comment by julien51** [comment]
@wawhal Any blocker that we can remove for you? Please let us know if you are stuck.

**Comment by wawhal** [comment]
Hey guys!
I am on vacation right now. My bad, I should have informed earlier.
I'll be back on the 2nd Jan 2018. Please wait if you can or you can go ahead and modify the PR.
Apologies.

**Comment by julien51** [comment]
Thanks @wawhal! Happy new year! Please let us know if you need anything else!

**Comment by wawhal** [comment]
Sorry for the delay folks. Will finish everything by 5th noon. Its been a while since this is open :D

**Comment by wawhal** [comment]
Hello folks. I am unable to perform local development.
Every `build`, `dev`, `start` command is failing with an error of types `Cannot resolve ... in ....`
Can you tell me if anything has changed in local development? Some additional step that we need to perform? 
@julien51 @cellog

**Comment by cellog** [comment]
Could you try running npm ci in root, smart-contracts, and unlock-app? That should reset your deps to pristine versions. At what stage do you see the error?

**Comment by cellog** [comment]
If you don't get this resolved by Monday, I can pull your branch and try to sort it out then

**Comment by wawhal** [comment]
@cellog I am unable to perform local development on master either.

This set of commands:

```
git clone git@github.com:unlock-protocol/unlock && cd unlock
npm ci
cd smart-contracts && npm ci
cd ../unlock-app && npm ci && npm run dev
```

fails with an error:

```
Saving artifacts...

> unlock-app@0.1.0 set-link /home/wawhal/oss/unlock/unlock-app
> run-script-os


> unlock-app@0.1.0 set-link:darwin:freebsd:linux:sunos /home/wawhal/oss/unlock/unlock-app
> pushd src && (test -e artifacts && rm -f artifacts) && ln -s ../../smart-contracts/build artifacts

sh: 1: pushd: not found
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! unlock-app@0.1.0 set-link:darwin:freebsd:linux:sunos: `pushd src && (test -e artifacts && rm -f artifacts) && ln -s ../../smart-contracts/build artifacts`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the unlock-app@0.1.0 set-link:darwin:freebsd:linux:sunos script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/wawhal/.npm/_logs/2019-01-05T17_41_28_377Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! unlock-app@0.1.0 set-link: `run-script-os`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the unlock-app@0.1.0 set-link script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/wawhal/.npm/_logs/2019-01-05T17_41_28_391Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! unlock-app@0.1.0 dev: `npm run setup-dev && npm run set-link && nodemon src/server.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the unlock-app@0.1.0 dev script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

**Comment by cellog** [comment]
I see the issue. Can you run bash or zsh? sh does not have pushd. I am going to remove that line it has caused too many troubles

**Comment by wawhal** [comment]
I am using bash @cellog
Anyways, I got it working temporarily by replacing `pushd` with `cd`.
Will push the changes in a bit.

**Comment by cellog** [comment]
Ok, good to know, what os are you running? Will be good to add to our machine matrix once we get one up and running

**Comment by wawhal** [comment]
I am on Ubuntu 16.04

**Comment by cellog** [comment]
P s. If you feel like doing a 3 character PR to replace pushd with cd that would be awesome, just make sure we end up back in the right direction with a cd .. at the end of the line. If not, I'll do it

File: unlock-app/src/components/creator/lock/KeyList.js (hunk 12 above)
Comment (wawhal) [reply unlock-app/src/components/creator/lock/KeyList.js]: It does not feel semantically correct to do that. We can't mix up the table with pagination component. You might want to add more customizations to `Table` or `KeyListWrapper` in future as they serve different functionalities.

**Comment by wawhal** [comment]
@cellog sure after this is merged and closed. The build is failing for some reasons. Everything works locally.

**Comment by julien51** [comment]
@wawhal I *think* you should be able to rebase and push again and the tests would run correctly this time. I am very very sorry about the issues you encountered. Please accept my apology.
We're very thankful for your contributions...

**Comment by wawhal** [comment]
@julien51 hey no issues. These are common problems :)
There's a bug in one of my recent commits. Will push a fix in US time. Sorry for more delay.

**Comment by julien51** [comment]
Haaaa, I see that the netlify deploy *also* fails when missing dependencies. Let me add another PR. Sorry about that, again!

**Comment by julien51** [comment]
One more rebase please?

**Comment by wawhal** [comment]
@julien51 are you sure it is the problem with the build system and not my PR? Locally everything seems to be working, the tests, storybook, UI etc. Is there anything else I should be adding/removing or checking as such?

**Comment by julien51** [comment]
cc @akeem It looks like the PR is failing due to tests in locksmith?

**Comment by akeem** [comment]
Hi @wawhal , my apologies for this. The fault should now be resolved in the latest master, please pull in the latest changes

**Comment by julien51** [comment]
Thanks @akeem ! 
@wawhal I just did a test by submitting an external PR (#951) and tests passed.

Please rebase and push again. This time should be good!
Thanks and sorry again for the trouble.

**Comment by julien51** [comment]
@cellog Please do a final check for @wawhal ;)

File: unlock-app/src/__tests__/components/interface/PaginationPageNumbers.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
+import React from 'react'
+import * as rtl from 'react-testing-library'
+
+import PageNumbers from '../../../components/interface/pagination/PageNumbers'
+
+jest.mock('next/link', () => {
+  return ({ children }) => children
+})
+
+afterEach(rtl.cleanup)
```
Comment (cellog) [inline unlock-app/src/__tests__/components/interface/PaginationPageNumbers.js]:
```suggestion
```

This is automatically done in our jest setup

File: unlock-app/src/stories/interface/PaginationPageNumbers.stories.js [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+      status: 'submitted',
+      confirmations: 0,
+      withdrawal: 'withdrawalsubmittedaddress',
+    },
+  },
+  keys: {
+    keyid: {
+      transaction: '0x23749328748932748932473298473289473298',
+      lock: '0xab7c74abc0c4d48d1bdad5dcb26153fc8780f83e',
+      expiration: Math.floor(new Date().getTime() / 1000) + 86400 * 30, // 30 days from right now
```
Comment (cellog) [inline unlock-app/src/stories/interface/PaginationPageNumbers.stories.js]: This needs to be changed, the snapshot tests will fail every new day :). You can see in the snapshots that "Feb 7, 2019" exists in the markup for every key, and that today, the snapshots failed and had to be updated to "Feb 8, 2019" Tomorrow they will need another update. Just hard-code a date here, it's fine for storybook purposes

File: unlock-app/src/stories/creator/KeyList.stories.js [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+  Object.keys(sampleLocks).forEach((numberStr, i) => {
+    const numberOfKeys = parseInt(numberStr, 10)
+    Array(numberOfKeys)
+      .fill()
+      .forEach((_, j) => {
+        keys[`${numberStr} ${j}`] = {
+          id: `${numberStr} ${j}`,
+          transaction: `0x23749328748932748932473298473289473298${i}${j}`,
+          lock: `0xab7c74abc0c4d48d1bdad5dcb26153fc8780f83e${numberStr}`,
+          expiration: Math.floor(new Date().getTime() / 1000) + 86400 * 30, // 30 days from right now
```
Comment (cellog) [inline unlock-app/src/stories/creator/KeyList.stories.js]: same comment, please hard-code a date here

File: unlock-app/src/components/interface/pagination/Pagination.js [hunk 20]
```diff
@@ -0,0 +1,130 @@
+import React from 'react'
+import styled from 'styled-components'
+import PreviousButtons from './Previous'
+import NextButtons from './Next'
+import PageNumbers from './PageNumbers'
+import { PGN_ITEMS_PER_PAGE } from '../../../constants'
+import UnlockPropTypes from '../../../propTypes'
```
Comment (cellog) [inline unlock-app/src/components/interface/pagination/Pagination.js]:
```suggestion
import UnlockPropTypes from '../../../propTypes'
import PropTypes from 'prop-types'
```

See below

File: unlock-app/src/components/interface/pagination/Pagination.js [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+class Paginate extends React.Component {
+  static propTypes = {
+    items: UnlockPropTypes.items,
+    renderItems: UnlockPropTypes.method,
```
Comment (cellog) [inline unlock-app/src/components/interface/pagination/Pagination.js]:
```suggestion
    renderItems: PropTypes.func,
```

File: unlock-app/src/propTypes.js [hunk 22]
```diff
@@ -66,6 +66,9 @@ export const element = PropTypes.oneOfType([PropTypes.func, PropTypes.element])
 
 export const delay = PropTypes.number
 
+export const items = PropTypes.array
+export const method = PropTypes.func
```
Comment (cellog) [inline unlock-app/src/propTypes.js]:
```suggestion
```

remove in favor of just using prop-types directly

File: unlock-app/src/propTypes.js [hunk 23]
```diff
@@ -75,11 +78,13 @@ export default {
   conversion,
   delay,
   element,
+  items,
   layout,
   lock,
   locks,
   key,
   keys,
+  method,
```
Comment (cellog) [inline unlock-app/src/propTypes.js]:
```suggestion
```

File: unlock-app/src/propTypes.js [hunk 24]
```diff
@@ -66,6 +66,9 @@ export const element = PropTypes.oneOfType([PropTypes.func, PropTypes.element])
 
 export const delay = PropTypes.number
 
+export const items = PropTypes.array
```
Comment (cellog) [inline unlock-app/src/propTypes.js]:
```suggestion
export const keyList = PropTypes.arrayOf(key)
```

File: unlock-app/src/propTypes.js [hunk 25]
```diff
@@ -75,11 +78,13 @@ export default {
   conversion,
   delay,
   element,
+  items,
```
Comment (cellog) [inline unlock-app/src/propTypes.js]:
```suggestion
  keyList,
```

File: unlock-app/src/components/interface/pagination/Pagination.js [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+import styled from 'styled-components'
+import PreviousButtons from './Previous'
+import NextButtons from './Next'
+import PageNumbers from './PageNumbers'
+import { PGN_ITEMS_PER_PAGE } from '../../../constants'
+import UnlockPropTypes from '../../../propTypes'
+
+class Paginate extends React.Component {
+  static propTypes = {
+    items: UnlockPropTypes.items,
```
Comment (cellog) [inline unlock-app/src/components/interface/pagination/Pagination.js]:
```suggestion
    items: UnlockPropTypes.keyList,
```

File: unlock-app/src/stories/creator/KeyList.stories.js [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+const addSampleKeys = () => {
+  const keys = {}
+  Object.keys(sampleLocks).forEach((numberStr, i) => {
+    const numberOfKeys = parseInt(numberStr, 10)
+    Array(numberOfKeys)
+      .fill()
+      .forEach((_, j) => {
+        keys[`${numberStr} ${j}`] = {
+          id: `${numberStr} ${j}`,
+          transaction: `0x23749328748932748932473298473289473298${i}${j}`,
```
Comment (cellog) [inline unlock-app/src/stories/creator/KeyList.stories.js]:
```suggestion
          transaction: `0x${i}${j}23749328748932748932473298473289473298`,
```

The end of the key is cut off so we can't actually see that the pagination works when clicking through pages

File: unlock-app/src/propTypes.js (hunk 24 above)
Comment (wawhal) [reply unlock-app/src/propTypes.js]: This is not correct. If the pagination component is reused somewhere else in the future, it could be an array of something else.

Comment (cellog) [reply unlock-app/src/propTypes.js]: We will update it then, go ahead with this for now 👍, as we don't have plans for pagination of anything else yet.

**Comment by wawhal** [comment]
We should be good to go @cellog ?

File: unlock-app/src/artifacts [hunk 28]
```diff
@@ -1 +1 @@
-../../smart-contracts/build/
\ No newline at end of file
+../../smart-contracts/build
```
[the remark this answers is not in the document]
Comment (julien51) [reply unlock-app/src/artifacts]: I *think* we can remove from git, so sending a PR for this now.

**Comment by hatgit** [comment]
Hi @julien51, and @cellog, following recent comments from the work by @wawhal, this is just a friendly reminder that if the work has been accepted, kindly login to your Gitcoin dashboard for this bounty where you can accept the work and pay out the bounty if you haven't already. Thanks!
